File: benchmarks/zig/b7_concurrent.zig
const std = @import("std");
const NUM_THREADS = 8;
const OPS = 10_000_000;
pub fn main(init: std.process.Init) !void {
    const io = init.io;
    var counter = std.atomic.Value(i64).init(0);
    var threads: [NUM_THREADS]std.Thread = undefined;
    for (&threads) |*t| {
        t.* = try std.Thread.spawn(.{}, worker, .{&counter});
    }
    for (threads) |t| t.join();
    var stdout_buffer: [256]u8 = undefined;
    var stdout_writer = std.Io.File.stdout().writer(io, &stdout_buffer);
    const stdout = &stdout_writer.interface;
    try stdout.print("{d}\n", .{counter.load(.seq_cst)});
    try stdout.flush();
}
fn worker(counter: *std.atomic.Value(i64)) void {
    for (0..OPS) |_| { _ = counter.fetchAdd(1, .monotonic); }
}

File: benchmarks/cpp/b7_concurrent.cpp
#include <iostream>
#include <thread>
#include <atomic>
#include <vector>
int main() {
    std::atomic<long long> counter{0};
    constexpr int T = 8, OPS = 10000000;
    std::vector<std::thread> threads;
    for (int i = 0; i < T; i++)
        threads.emplace_back([&]{ for (int j = 0; j < OPS; j++) counter.fetch_add(1, std::memory_order_relaxed); });
    for (auto& t : threads) t.join();
    std::cout << counter.load() << "\n";
}
